protonvpn_cli/connection.py

```diff
@@ -639,55 +639,68 @@ def manage_ipv6(mode):
     """
 
     ipv6_backupfile = os.path.join(CONFIG_DIR, "ipv6.backup")
+    ip6tables_backupfile = os.path.join(CONFIG_DIR, "ip6tables.backup")
 
     if mode == "disable":
 
         logger.debug("Disabling IPv6")
+        # Needs to be removed eventually. I'll leave it in for now
+        # so it still properly restores the IPv6 address the old way
         if os.path.isfile(ipv6_backupfile):
-            manage_ipv6("restore")
-
-        default_route = subprocess.run(
-            "ip route show | grep default",
-            stdout=subprocess.PIPE, shell=True
-        )
-
-        # Get the default nic from ip route show output
-        default_nic = default_route.stdout.decode().strip().split()[4]
-
-        ipv6_info = subprocess.run(
-            "ip addr show dev {0} | grep '\<inet6.*global\>'".format(default_nic), # noqa
-            stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True,
-        )
+            manage_ipv6("legacy_restore")
 
-        has_ipv6 = True if ipv6_info.returncode == 0 else False
+        if os.path.isfile(ip6tables_backupfile):
+            logger.debug("IPv6 backup exists")
+            manage_ipv6("restore")
 
-        # Stop configuration if not IPv6 address is present
-        if not has_ipv6:
-            logger.debug("No IPv6 present")
-            return
+        # Backing up ip6ables rules
+        logger.debug("Backing up ip6tables rules")
+        ip6tables_rules = subprocess.run(["ip6tables-save"],
+                                        stdout=subprocess.PIPE)
 
-        # Get the actual IPv6 address
-        ipv6_addr = ipv6_info.stdout.decode().strip().split()[1]
+        if "COMMIT" in ip6tables_rules.stdout.decode():
```

Comment:
What's the decode here? What happens if there are some non-ascii characters in the firewall?

Reply:
subprocess returns a bytestring. To compare it with a python string (unicode, I think), I need to decode the output from subprocess, turning it into a normal string. Or am I misunderstanding.

Reply:
I have not yet upgraded my brain to Python3, so wanted just to be sure :+1: